src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -648,6 +655,132 @@ public byte[] Http2AltSvcOriginUri
             return await GetHttpConnectionAsync(request, async, cancellationToken).ConfigureAwait(false);
         }
 
+        private Http2Connection? FindHttp2ConnectionWithFreeStreamSlot(Http2Connection[]? connections)
+        {
+            if (connections == null)
+            {
+                return null;
+            }
+
+            // Try finding an existing connection with available stream's quota.
+            for (int i = 0; i < connections.Length; i++)
+            {
+                if (connections[i].AcquireStreamSlot())
+                {
+                    if (NetEventSource.Log.IsEnabled())
+                    {
+                        Trace("Using existing HTTP2 connection.");
+                    }
+
+                    return connections[i];
+                }
+            }
+            return null;
+        }
+
+        private async Task<Http2Connection?> GetHttp2ConnectionAcceptingNewStreams()
+        {
+            if (!_poolManager.Settings.EnableMultipleHttp2Connections)
+            {
+                // Fast shortcut for the most common case.
+                Http2Connection[]? localConnections = _http2Connections;
+                return localConnections != null && IsAlive(localConnections[0]) ? localConnections[0] : null;
+            }
+
+            while (true)
+            {
+                Http2Connection[]? localConnections = _http2Connections;
+
+                if (localConnections == null)
+                {
+                    return null;
+                }
+
+                for (int i = 0; i < localConnections.Length; i++)
+                {
+                    Http2Connection http2Connection = localConnections[i];
+
+                    if (IsAlive(http2Connection) && http2Connection.AcquireStreamSlot())
+                    {
+                        return http2Connection;
+                    }
+                }
+
+                // No connection with a free stream slot found.
+
+                if (localConnections.Length < _poolManager.Settings._maxHttp2ConnectionsPerServer)
+                {
+                    // Procceed to a new connection construction.
+                    return null;
+                }
+
+                // Connection maximum is reached. Wait for a free slot on any existing connection.
+
+                SemaphoreSlim? localfreeConnections;
+                lock (SyncObj)
+                {
+                    _waitingRequestCount++;
+                    if (_freeStreamSlots == null)
+                    {
+                        _freeStreamSlots = new SemaphoreSlim(0);
+                    }
+                    localfreeConnections = _freeStreamSlots;
+                }
+
+                await localfreeConnections.WaitAsync().ConfigureAwait(false);
```

Comment:
Where is _freeStreamSlots updated if, say, a connection has its stream limit raised?

Reply:
It was missing, but fixed now. Thanks!